--- a/src/module/chttrans/chttrans.c
+++ b/src/module/chttrans/chttrans.c
@@ -71,6 +71,8 @@
 CONFIG_BINDING_REGISTER("TraditionalChinese", "Hotkey", hkToggle)
 CONFIG_BINDING_REGISTER_WITH_FILTER("TraditionalChinese", "EnabledForIM",
                                     strEnableForIM, ChttransEnabledForIMFilter)
+CONFIG_BINDING_REGISTER("TraditionalChinese", "ConfigSimptoTrad", cfgs2t)
+CONFIG_BINDING_REGISTER("TraditionalChinese", "ConfigTradtoSimp", cfgt2s)
 CONFIG_BINDING_END()
 
 FCITX_DEFINE_PLUGIN(fcitx_chttrans, module, FcitxModule) = {
--- a/src/module/chttrans/chttrans-opencc.c
+++ b/src/module/chttrans/chttrans-opencc.c
@@ -47,8 +47,12 @@
 
     if (!OpenCCLoadLib())
         return false;
-    transState->ods2t = _opencc_open(_OPENCC_DEFAULT_CONFIG_SIMP_TO_TRAD);
-    transState->odt2s = _opencc_open(_OPENCC_DEFAULT_CONFIG_TRAD_TO_SIMP);
+    transState->ods2t = _opencc_open(transState->cfgs2t);
+    transState->odt2s = _opencc_open(transState->cfgt2s);
+    if (transState->ods2t == -1)
+        transState->ods2t = _opencc_open(_OPENCC_DEFAULT_CONFIG_SIMP_TO_TRAD);
+    if (transState->odt2s == -1) 
+        transState->odt2s = _opencc_open(_OPENCC_DEFAULT_CONFIG_TRAD_TO_SIMP);
     if (!transState->ods2t && !transState->odt2s)
         return false;
     return true;
--- a/src/module/chttrans/chttrans_p.h
+++ b/src/module/chttrans/chttrans_p.h
@@ -32,6 +32,8 @@
     void* odt2s;
     FcitxInstance* owner;
     boolean openccLoaded;
+    char* cfgs2t;
+    char* cfgt2s;
 } FcitxChttrans;
 
 #endif // CHTTRANS_P_H
--- a/src/module/chttrans/fcitx-chttrans.desc
+++ b/src/module/chttrans/fcitx-chttrans.desc
@@ -17,5 +17,17 @@
 Description=Enabled state for im
 Advance=True
 
+[TraditionalChinese/ConfigSimptoTrad]
+Type=String
+DefaultValue=zhs2zht.ini
+Description=Configuration for Conversion from Simplified Chinese to Traditional Chinese
+Advance=True
+
+[TraditionalChinese/ConfigTradtoSimp]
+Type=String
+DefaultValue=zht2zhs.ini
+Description=Configuration for Conversion from Traditional Chinese to Simplified Chinese
+Advance=True
+
 [DescriptionFile]
 LocaleDomain=fcitx
