From 40c3c25e1b0e4c833e3f3a322c49c26133dfdfef Wed Jul 24 12:43:00 2013
From: Weng Xuetian <wengxt@gmail.com>
Date: Wed Jul 24 20:43:00 CST 2013
Subject: [PATCH] [rime] setup logging

--- a/src/fcitx-rime.c
+++ b/src/fcitx-rime.c
@@ -69,6 +69,7 @@
 {
     FcitxRime* rime = (FcitxRime*) fcitx_utils_malloc0(sizeof(FcitxRime));
     rime->owner = instance;
+    RimeSetupLogging("rime.fcitx-rime");
     FcitxRimeStart(rime, false);
 
     FcitxIMIFace iface;
