--- a/src/fcitx-rime.c
+++ b/src/fcitx-rime.c
@@ -254,6 +254,11 @@
 
 void FcitxRimeUpdateStatus(FcitxRime* rime)
 {
+    if (RimeIsMaintenancing())
+        return;
+    if (!RimeFindSession(rime->session_id)) {
+        rime->session_id = RimeCreateSession();
+    }
     RimeStatus status = {0};
     RIME_STRUCT_INIT(RimeStatus, status);
     if (RimeGetStatus(rime->session_id, &status)) {
@@ -513,7 +518,12 @@
 
 static const char* FcitxRimeGetIMIcon(void* arg)
 {
+    if (RimeIsMaintenancing())
+        return "@rime-disabled";
     FcitxRime* rime = (FcitxRime*) arg;
+    if (!RimeFindSession(rime->session_id)) {
+        rime->session_id = RimeCreateSession();
+    }
     RimeStatus status = {0};
     RIME_STRUCT_INIT(RimeStatus, status);
     if (RimeGetStatus(rime->session_id, &status)) {
@@ -548,7 +558,12 @@
 
 static const char* FcitxRimeGetChtsIcon(void *arg)
 {
+    if (RimeIsMaintenancing())
+        return "@rime-chtrad";
     FcitxRime* rime = (FcitxRime*) arg;
+    if (!RimeFindSession(rime->session_id)) {
+        rime->session_id = RimeCreateSession();
+    }
     if (rime->session_id) {
         if(RimeGetOption(rime->session_id, "simplification")) {
             return "rime-chsimp";
