--- a/src/fcitx-rime.c
+++ b/src/fcitx-rime.c
@@ -270,6 +270,9 @@
 
 void FcitxRimeUpdateStatus(FcitxRime* rime)
 {
+    if (!RimeFindSession(rime->session_id)) {
+        rime->session_id = RimeCreateSession();
+    }
     RimeStatus status = {0};
     RIME_STRUCT_INIT(RimeStatus, status);
     if (RimeGetStatus(rime->session_id, &status)) {
@@ -530,6 +533,9 @@
 static const char* FcitxRimeGetIMIcon(void* arg)
 {
     FcitxRime* rime = (FcitxRime*) arg;
+    if (!RimeFindSession(rime->session_id)) {
+        rime->session_id = RimeCreateSession();
+    }
     RimeStatus status = {0};
     RIME_STRUCT_INIT(RimeStatus, status);
     if (RimeGetStatus(rime->session_id, &status)) {
