--- a/src/fcitx-rime.c
+++ b/src/fcitx-rime.c
@@ -3,6 +3,7 @@
 #include <fcitx/candidate.h>
 #include <fcitx/hook.h>
 #include <fcitx-config/xdg.h>
+#include <fcitx/module/freedesktop-notify/fcitx-freedesktop-notify.h>
 #include <libintl.h>
 #include <rime_api.h>
 
@@ -46,6 +47,58 @@
 FCITX_EXPORT_API
 int ABI_VERSION = FCITX_ABI_VERSION;
 
+void FcitxRimeNotificationHandler(void* context_object,
+                                  RimeSessionId session_id,
+                                  const char* message_type,
+                                  const char* message_value)
+{
+    FcitxRime* rime = (FcitxRime*) context_object;
+    char* message = NULL;
+    const char* staticon = NULL;
+
+    if (!strcmp(message_type, "deploy")) {
+        staticon = "fcitx-rime-deploy";
+        if (!strcmp(message_value, "start")) {
+            message = _("Rime is under maintenance ...");
+        } else if (!strcmp(message_value, "success")) {
+            message = _("Rime is ready.");
+        } else if (!strcmp(message_value, "failure")) {
+            message = _("Rime has encountered an error. "
+                        "See /tmp/rime.fcitx.ERROR for details.");
+        }
+    }
+
+    if (!strcmp(message_type, "option")) {
+        if (!strcmp(message_value, "ascii_mode")) {
+            message = _("Switch to English mode.");
+            staticon = "rime-latin";
+        } else if (!strcmp(message_value, "full_shape")) {
+            message = _("Switch to full-width mode.");
+            staticon = "fcitx-fullwidth-active";
+        } else if (!strcmp(message_value, "!full_shape")) {
+            message = _("Switch to half-width mode.");
+            staticon = "fcitx-fullwidth-inactive";
+        } else if (!strcmp(message_value, "simplification")) {
+            message = _("Enable Traditional Chinese to Simplified Chinese.");
+        } else if (!strcmp(message_value, "!simplification")) {
+            message = _("Disable Traditional Chinese to Simplified Chinese.");
+       }
+    }
+
+    if (!strcmp(message_type, "schema")) {
+        fcitx_utils_alloc_cat_str(message, _("Selected schema: "),
+                                  message_value);
+        staticon = "rime-im";
+    }
+
+    if (message)
+        FcitxFreeDesktopNotifyShowAddonTip(rime->owner, "fcitx-rime",
+                                           staticon, _("Rime"), message);
+
+    if (!strcmp(message_type, "schema"))
+        fcitx_utils_free(message);
+}
+
 static void FcitxRimeStart(FcitxRime* rime, boolean fullcheck) {
 
     char* user_path = NULL;
@@ -56,6 +109,8 @@
     //char* shared_data_dir = fcitx_utils_get_fcitx_path_with_filename("pkgdatadir", "rime");
     const char* shared_data_dir = RIME_DATA_DIR;
 
+    RimeSetNotificationHandler(FcitxRimeNotificationHandler, rime);
+
     RimeTraits fcitx_rime_traits;
     fcitx_rime_traits.shared_data_dir = shared_data_dir;
     fcitx_rime_traits.user_data_dir = user_path;
